### PR TITLE
Ensure slack users are active members

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

See https://github.com/dust-tt/decisions/issues/153.

Remake of https://github.com/dust-tt/dust/pull/3565.

This PR introduces a new layer to our Slackbot, restricting the usage of our Slack integrations to active members only.  The goal here is to prevent all users of a Slack workspace to use Dust without having an account.

This rules apply for everyone except for:
- Whitelisted chatbots.
- Dust workspaces with a Slack whitelisted domains.

## Key updates in this PR include:
- Refactoring the existing logic of guest users in a Slack workspace.
- Expose the list of all active members in a Dust workspace through the front API. To avoid doing API calls on every slack messages we're caching these results for 15 minutes.
- Expose the list of verified domains for a workspace through the front API. To avoid doing API calls on every slack messages we're caching these results for 15 minutes.
- Returning a slack message to invite users to sign up on Dust based on the workspace settings.
- A mechanism to safely release. In case things go wrong, using our CLI we will be able to circumvent the new implementation.
```
./admin/cli.sh slack whitelist-domains --wId <xxx> --whitelistedDomains <yyy>
```

Two scenarios:

1. The workspace has auto join enabled, the user gets:
image

2. The workspace does not have auto join enabled, the user get:
image

## Risk

Pretty risky, tested locally with many users and different scenarios. Rollout system in place in case we need to course correct. Safe to rollback.

## Deploy Plan

Given the dependency on the Front api to get the count of active members within the workspace, we need to first deploy front and then deploy connectors.

### CODE REVIEW

File: connectors/src/connectors/slack/lib/slack_client.ts [hunk 1]
```diff
@@ -109,62 +110,13 @@ export async function getSlackUserInfo(slackClient: WebClient, userId: string) {
...
 
-async function getSlackConversationInfo(
+export async function getSlackConversationInfo(
   slackClient: WebClient,
   channelId: string
 ) {
   return slackClient.conversations.info({ channel: channelId });
 }
 
-// Verify the Slack user is not an external guest to the workspace.
```
Comment (flvndvd) [inline connectors/src/connectors/slack/lib/slack_client.ts:119]: Moved to workspace_limit.ts

File: connectors/src/connectors/slack/lib/workspace_limits.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        },
+      ],
+    },
+  ],
+  autojoin_disabled: [
+    {
+      type: "section",
+      text: {
+        type: "mrkdwn",
+        text: "Oops! It looks like you're not a member of Dust yet. Please reach out to a Dust admin to join the workspace and start using the Slack integration.",
```
Comment (spolu) [inline connectors/src/connectors/slack/lib/workspace_limits.ts:146]: Let's drop the Ooops ?

Let's review wording with @gabhubert

**Review by spolu — COMMENTED** [review]
Looks good. 
Authentication fixes required :+1:

File: connectors/src/connectors/slack/lib/workspace_limits.ts (hunk 2 above)
Comment (flvndvd) [reply connectors/src/connectors/slack/lib/workspace_limits.ts:146]: Will check the copy with @gabhubert.

File: front/pages/api/v1/w/[wId]/verified_domains.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      api_error: {
+        type: "workspace_not_found",
+        message: "The workspace was not found.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "GET":
+      const verifiedDomain = await getWorkspaceVerifiedDomain(owner.id);
```
[the remark this answers is not in the document]
Comment (flvndvd) [reply front/pages/api/v1/w/[wId]/verified_domains.ts]: Fixed.

File: front/pages/api/v1/w/[wId]/members/emails.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  if (keyRes.isErr()) {
+    return apiError(req, res, keyRes.error);
+  }
+  const { auth } = await Authenticator.fromKey(
+    keyRes.value,
+    req.query.wId as string
+  );
+
+  const owner = auth.workspace();
+  const isSystemKey = keyRes.value.isSystem;
```
[the remark this answers is not in the document]
Comment (flvndvd) [reply front/pages/api/v1/w/[wId]/members/emails.ts:26]: Fixed.

File: front/pages/api/v1/w/[wId]/verified_domains.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  if (keyRes.isErr()) {
+    return apiError(req, res, keyRes.error);
+  }
+  const { auth } = await Authenticator.fromKey(
+    keyRes.value,
+    req.query.wId as string
+  );
+
+  const owner = auth.workspace();
+  const isSystemKey = keyRes.value.isSystem;
```
[the remark this answers is not in the document]
Comment (flvndvd) [reply front/pages/api/v1/w/[wId]/verified_domains.ts:26]: Fixed.

File: connectors/src/connectors/slack/lib/workspace_limits.ts [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+  return isDomainAutoJoinEnabled?.domainAutoJoinEnabled ?? false;
+}
+
+const slackMembershipAccessBlocks = {
+  autojoin_enabled: [
+    {
+      type: "section",
+      text: {
+        type: "mrkdwn",
+        text: "Access the Slack integration by joining your Dust workspace. Only members can use this feature. Click 'Join My Workspace' to get started. For help, contact a Dust admin.",
```
Comment (gabhubert) [inline connectors/src/connectors/slack/lib/workspace_limits.ts:120]: The Slack integration is accessible to members of your company's Dust workspace. Click 'Join My Workspace' to get started. For help, contact an administrator.

File: connectors/src/connectors/slack/lib/workspace_limits.ts (hunk 2 above)
Comment (gabhubert) [inline connectors/src/connectors/slack/lib/workspace_limits.ts:146]: It looks like you're not a member of your company's Dust workspace yet. Please reach out to an administrator to join and start using Dust on Slack.

**Review by gabhubert — COMMENTED** [review]
left comments

File: connectors/src/connectors/slack/lib/workspace_limits.ts (hunk 6 above)
Comment (flvndvd) [reply connectors/src/connectors/slack/lib/workspace_limits.ts:120]: Fixed.

File: connectors/src/connectors/slack/lib/workspace_limits.ts (hunk 2 above)
Comment (flvndvd) [reply connectors/src/connectors/slack/lib/workspace_limits.ts:146]: Fixed.

**Review by spolu — APPROVED** [review]

**Comment by spolu** [comment]
:money_with_wings:

**Pull request merged**